### PR TITLE
Implement user-to-user chat function and messages inbox

### PR DESCRIPTION
**Pull request by MaximParker**
- Adds 'Messages' route which gets the user's conversations from Firebase
- Adds the 'Chat' component which renders an individual conversation and allows the user to send messages
- Adds the sendWelcomeMessage function which establishes the folders for a user's messages when they register
- Adds the sendMessage function
- Adds displayName attribute to User object in stores.js when logging in

NB: In order to chat between two users, the startChatroom function must be instigated by one of those users using the pair's uids at some prior point.

### CODE REVIEW

File: src/components/Chat.svelte [hunk 1]
```diff
@@ -1,52 +1,96 @@
...
+    on:submit={() => {
+      currentChat = "";
+      navigate("/messages");
+    }}
+  >
+    <button type="submit">Go back to messages</button>
+  </form>
+
+  <h1>{currentChat}</h1>
+  {console.log("Current chat:", currentChat)}
```
Comment (dweverson) [inline src/components/Chat.svelte:68]: console log left in

File: src/components/Chat.svelte [hunk 2]
```diff
@@ -1,52 +1,96 @@
...
+    );
+    console.log("Sending message: ", senderCopy.id, "(sender's copy)");
+    const recipientCopy = await addDoc(
+      collection(
+        db,
+        `messages/${currentChat}/conversations/${signedIn.uid}/messages`
+      ),
+      newMessage
+    );
+    console.log("Sending message: ", recipientCopy.id, "(recipient's copy)");
```
Comment (dweverson) [inline src/components/Chat.svelte:52]: console.log left in

File: src/components/Chat.svelte [hunk 3]
```diff
@@ -1,52 +1,96 @@
...
-  }
+  const sendMessage = async () => {
+    const senderCopy = await addDoc(
+      collection(
+        db,
+        `messages/${signedIn.uid}/conversations/${currentChat}/messages`
+      ),
+      newMessage
+    );
+    console.log("Sending message: ", senderCopy.id, "(sender's copy)");
```
Comment (dweverson) [inline src/components/Chat.svelte:44]: console.log left in

File: src/utils/api.js [hunk 4]
```diff
@@ -74,18 +80,97 @@ export const reseedListingsDatabase = async (event, listings) => {
...
+    location: "York",
+    geotag: "53.96, -1.09",
     tradeRequired: true,
   };
   postListing(event);
   console.log("Re-seed complete.");
 };
+
+export const sendWelcomeMessage = (targetID) => {
+  console.log(`Creating folder in messages for ${targetID}`);
```
Comment (dweverson) [inline src/utils/api.js:102]: console.log left in

File: src/utils/api.js [hunk 5]
```diff
@@ -74,18 +80,97 @@ export const reseedListingsDatabase = async (event, listings) => {
...
+  return setDoc(doc(db, `messages`, `${targetID}`), {}).then(() => {
+    console.log(`Creating conversations collection for ${targetID}`);
```
Comment (dweverson) [inline src/utils/api.js:104]: console.log left in

File: src/utils/api.js [hunk 6]
```diff
@@ -74,18 +80,97 @@ export const reseedListingsDatabase = async (event, listings) => {
...
+          text: "Welcome to Topswop! Look here for your messages.",
+          read: false,
+        }
+      );
+    });
+  });
+};
+
+export const createChatroom = (uid_a, uid_b, displayName_a, displayname_b) => {
+  console.log(`Creating a conversation between ${uid_a} and ${uid_b}`);
```
Comment (dweverson) [inline src/utils/api.js:126]: console.log left in

File: src/utils/api.js [hunk 7]
```diff
@@ -74,18 +80,97 @@ export const reseedListingsDatabase = async (event, listings) => {
...
+  );
+};
+
+export const sendDirectMessage = (
+  sender_id,
+  sender_displayName,
+  recipient_id,
+  text
+) => {
+  console.log(`Creating conversations collection for ${targetID}`);
```
Comment (dweverson) [inline src/utils/api.js:158]: console.log left in

File: src/utils/auth.js [hunk 8]
```diff
@@ -1,36 +1,56 @@
...
 
 export const loginByEmail = (email, password) => {
   return signInWithEmailAndPassword(auth, email, password)
-  .then((userCredential) => {
-    console.log(userCredential.user);
-    user.set({
-      uid: userCredential.user.uid,
-      email: userCredential.user.email,
+    .then((userCredential) => {
+      console.log(userCredential);
```
Comment (dweverson) [inline src/utils/auth.js:17]: console.log left in

File: src/utils/auth.js [hunk 9]
```diff
@@ -1,36 +1,56 @@
...
+      alert(error);
+    });
 };
 
-export const registerUserByEmail = (email, password) => {
+export const registerUserByEmail = (email, password, username) => {
   return createUserWithEmailAndPassword(auth, email, password)
     .then((userCredential) => {
-      console.log(userCredential.user);
+      console.log("REGISTERED NEW USER:\n",userCredential.user);
```
Comment (dweverson) [inline src/utils/auth.js:47]: console.log left in

**Review by dweverson — CHANGES_REQUESTED** [review]
Looks good, have commented console logs left in but also understand if your keeping them for time being whilst still working on the feature.

**Comment by MaximParker** [comment]
Resolved removal of console logs

**Review by dweverson — APPROVED** [review]
as discussed console logs to be removed on next merge

**Pull request merged**